optaplanner-core/src/main/java/org/optaplanner/core/impl/constructionheuristic/DefaultConstructionHeuristicPhaseFactory.java

```diff
@@ -113,6 +108,78 @@ public DefaultConstructionHeuristicPhaseFactory(ConstructionHeuristicPhaseConfig
         return builder.build();
     }
 
+    private Optional<EntityPlacerConfig> getValidEntityPlacerConfig() {
+        return Optional.ofNullable(phaseConfig.getEntityPlacerConfig()).filter(entityPlacerConfig -> {
+            if (phaseConfig.getConstructionHeuristicType() != null) {
+                throw new IllegalArgumentException(
+                        "The constructionHeuristicType (" + phaseConfig.getConstructionHeuristicType()
+                                + ") must not be configured if the entityPlacerConfig (" + entityPlacerConfig
+                                + ") is explicitly configured.");
+            }
+            if (phaseConfig.getMoveSelectorConfigList() != null) {
+                throw new IllegalArgumentException("The moveSelectorConfigList (" + phaseConfig.getMoveSelectorConfigList()
+                        + ") cannot be configured if the entityPlacerConfig (" + entityPlacerConfig
+                        + ") is explicitly configured.");
+            }
+            return true;
+        });
```

Comment:
To me, this is misuse of `Optional`. (And also bikeshedding. :-)) You are trying to avoid `null`, but there is nothing wrong with `null`. It never escapes the method, therefore it is safe.

    private Optional<EntityPlacerConfig> getValidEntityPlacerConfig() {
        EntityPlacerConfig entityPlacerConfig = phaseConfig.getEntityPlacerConfig();
        if (entityPlacerConfig == null) {
            return Optional.empty();
        }
        if (phaseConfig.getConstructionHeuristicType() != null) {
            throw new IllegalArgumentException(
                    "The constructionHeuristicType (" + phaseConfig.getConstructionHeuristicType()
                            + ") must not be configured if the entityPlacerConfig (" + entityPlacerConfig
                            + ") is explicitly configured.");
        }
        if (phaseConfig.getMoveSelectorConfigList() != null) {
            throw new IllegalArgumentException("The moveSelectorConfigList (" + phaseConfig.getMoveSelectorConfigList()
                    + ") cannot be configured if the entityPlacerConfig (" + entityPlacerConfig
                    + ") is explicitly configured.");
        }
        return Optional.of(entityPlacerConfig);
    }

Doesn't this read better?

Reply:
Yes, I agree it reads better and I'm going to apply your suggestion.